Beams/MRML/vtkMRMLRTIonBeamNode.cxx

```diff
@@ -0,0 +1,514 @@
+
+  See COPYRIGHT.txt
+  or http://www.slicer.org/copyright/copyright.txt for details.
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  This file was originally developed by Kevin Wang, Princess Margaret Cancer Centre
+  and was supported by Cancer Care Ontario (CCO)'s ACRU program
+  with funds provided by the Ontario Ministry of Health and Long-Term Care
+  and Ontario Consortium for Adaptive Interventions in Radiation Oncology (OCAIRO).
+
+==============================================================================*/
+
+// Beams includes
+#include "vtkMRMLRTIonBeamNode.h"
+#include "vtkMRMLRTPlanNode.h"
+
+// SlicerRT includes
+#include "vtkSlicerRtCommon.h"
+
+// MRML includes
+#include <vtkMRMLModelDisplayNode.h>
+#include <vtkMRMLScalarVolumeNode.h>
+#include <vtkMRMLDoubleArrayNode.h>
+#include <vtkMRMLTableNode.h>
+#include <vtkMRMLMarkupsFiducialNode.h>
+#include <vtkMRMLLinearTransformNode.h>
+#include <vtkMRMLSubjectHierarchyNode.h>
+#include <vtkMRMLSubjectHierarchyConstants.h>
+
+// VTK includes
+#include <vtkCommand.h>
+#include <vtkObjectFactory.h>
+#include <vtkSmartPointer.h>
+#include <vtkIntArray.h>
+#include <vtkTransform.h>
+#include <vtkTransformPolyDataFilter.h>
+#include <vtkDoubleArray.h>
+#include <vtkTable.h>
+#include <vtkCellArray.h>
+
+//------------------------------------------------------------------------------
+namespace
+{
+
+const char* const SCANSPOT_REFERENCE_ROLE = "ScanSpotRef";
+constexpr double FWHM_TO_SIGMA = 1. / (2. * sqrt(2. * log(2.)));
+
+} // namespace
+
+//------------------------------------------------------------------------------
+vtkMRMLNodeNewMacro(vtkMRMLRTIonBeamNode);
+
+//----------------------------------------------------------------------------
+vtkMRMLRTIonBeamNode::vtkMRMLRTIonBeamNode()
+  :
+  Superclass(),
+  VSADx(Superclass::SAD),
+  IsocenterToJawsDistanceX(Superclass::SourceToJawsDistanceX),
+  IsocenterToJawsDistanceY(Superclass::SourceToJawsDistanceY),
+  IsocenterToMultiLeafCollimatorDistance(Superclass::SourceToMultiLeafCollimatorDistance),
+  IsocenterToRangeShifterDistance(4000.),
+  ScanningSpotSize({ 15.0, 15.0 })
+{
+  this->VSADx = 6500.0;
+  this->VSADy = 6500.0;
+  this->IsocenterToJawsDistanceX = 3000.;
+  this->IsocenterToJawsDistanceY = 3000.;
+  this->IsocenterToMultiLeafCollimatorDistance = 2500.;
+}
+
+//----------------------------------------------------------------------------
+vtkMRMLRTIonBeamNode::~vtkMRMLRTIonBeamNode()
+{
+  this->SetBeamDescription(nullptr);
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLRTIonBeamNode::WriteXML(ostream& of, int nIndent)

```

Comment:
Please use the new macros defined here
https://github.com/Slicer/Slicer/blob/master/Libs/MRML/Core/vtkMRMLNodePropertyMacros.h
See examples here
https://github.com/Slicer/Slicer/blob/master/Libs/MRML/Core/vtkMRMLDisplayNode.cxx#L110
Also for ReadXMLAttributes, Copy, and PrintSelf

Reply:
I fixed that, not so sure about Copy method though.